### ISSUE TITLE
UP010 should not report future features as unnecessary if they are used

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#19561]
### Summary

If a feature imported from `__future__` is used later in the module, it is not unnecessary, so [`unnecessary-future-import` (UP010)](https://docs.astral.sh/ruff/rules/unnecessary-future-import/) should not flag it. [Example](https://play.ruff.rs/d08da604-d150-4389-9c90-d3e727e12bdf):
```console
$ cat >up010.py <<'# EOF'
from __future__ import print_function
print(print_function)
# EOF

$ python up010.py
_Feature((2, 6, 0, 'alpha', 2), (3, 0, 0, 'alpha', 0), 1048576)

$ ruff --isolated check up010.py --select UP010 --fix
Found 1 error (1 fixed, 0 remaining).

$ python up010.py 2>&1 | tail -n 1
NameError: name 'print_function' is not defined
```

This also applies to reexports using `__all__` [like this](https://play.ruff.rs/025537c1-ecd4-4b57-a12c-1b13284a448e):
```python
from __future__ import print_function
__all__ = ["print_function"]
```

### Version

ruff 0.12.5 (d13228ab8 2025-07-24)

**Comment by mikeleppane** [comment]
Hi! I have reviewed this, and it is my understanding that the necessary adjustments/fixes may result in a somewhat wider blast radius.

**Comment by ntBre** [comment]
> Hi! I have reviewed this, and it is my understanding that the necessary adjustments/fixes may result in a somewhat wider blast radius.

Thanks for looking into it! Could you expand a bit on the wider blast radius? I assumed you meant that this would impact other rules (like changing the semantic model somehow), but from a quick peek at the rule implementation, it seems like it could be pretty localized, although it sounds like we may want to reuse some code from [unused-import (F401)](https://docs.astral.sh/ruff/rules/unused-import/#unused-import-f401).

I could be wrong, though. I'm interested in what you found!

**Comment by mikeleppane** [comment]
Yes, you are correct! The `unused-import` rule can address the cases I was looking for. However, the current implementation of `UP010` cannot accommodate the possible use cases.
